Prevent polling from recreating an entity after removal

bdraco opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
- There was a race where the polling could finish after the entity
  had been removed which would cause it to be unexpectedly recreated


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #66450
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

bdraco committed:
Prevent polling from recreating an entity after removal

- There was a race where the polling could finish after the entity
  had been removed which would cause it to be unexpectedly recreated

mib1185 commented:
Can confirm, that this works and fixes #66450

bdraco committed:
cover

thecode commented:
Not directly related, this reminded me that the coordinator would benefit from some improvements in this area also, currently it may also poll during unload or shutdown, we have some workarounds in Shelly to overcome this problem.

bdraco commented:
It looks like the failing tests are removing the entity before it completes its first update

bdraco committed:
rename to prevent conflict

bdraco commented:
Actually gdacs uses `self._removed` so I'll rename this `self._entity_removed`

mib1185 reviewed (approved):
LGTM 👍  thx @bdraco for this fast fix, just 10min after my issue description on discord 😃

thecode commented:
All tests pass now but to be honest I would prefer fixing the integration which uses the `self._removed` just to be consistent :-), Anyhow LGTM

bdraco commented:
> All tests pass now but to be honest I would prefer fixing the integration which uses the `self._removed` just to be consistent :-), Anyhow LGTM
```
homeassistant/components/agent_dvr/camera.py:        self._removed = False
homeassistant/components/agent_dvr/camera.py:            if self._removed:
homeassistant/components/agent_dvr/camera.py:            self._removed = False
homeassistant/components/agent_dvr/camera.py:            if self.device.client.is_available and not self._removed:
homeassistant/components/agent_dvr/camera.py:                self._removed = True
homeassistant/components/gdacs/sensor.py:        self._removed = None
homeassistant/components/gdacs/sensor.py:        self._removed = status_info.removed
homeassistant/components/gdacs/sensor.py:            (ATTR_REMOVED, self._removed),
homeassistant/components/geonetnz_quakes/sensor.py:        self._removed = None
homeassistant/components/geonetnz_quakes/sensor.py:        self._removed = status_info.removed
homeassistant/components/geonetnz_quakes/sensor.py:            (ATTR_REMOVED, self._removed),
```
Looks like there are a few cases plus what exists in custom land so I was worried about creating a breaking change.

mib1185 commented:
maybe we should (_in a later PR_) rename the internaly used properties to __xxxx to mark them as just internal used and avoid future conflicts with integrations 🤔

thecode commented:
> maybe we should (_in a later PR_) rename the internaly used properties to `__xxxx` to mark them as just internal used and avoid future conflicts with integrations 🤔

I also had that in mind, but this can happen in a separate code quality PR. we should also consider adding this PR to the next milestone as it fixes an important race situation.

chemelli74 reviewed (approved):
Thx Nick, as always!

bdraco commented on homeassistant/helpers/entity.py:
```diff
@@ -767,6 +773,7 @@ def add_to_platform_start(
         self.platform = platform
         self.parallel_updates = parallel_updates
         self._added = True
+        self._entity_removed = False
```
I was concerned that `add_to_platform_abort` set `self._added = False` and it might have an unexpected side effect if the entity was added later after that was called.  If thats not an issue, I think we can go with checking `_added`

bdraco replied:
Just tried that and it also appears to break `update_before_add`

bdraco replied:
We need a trinary state for this

bdraco replied:
I made it an enum instead.

bdraco committed:
use enum

bdraco committed:
naming

bdraco committed:
naming

bdraco committed:
cleanup comparison

balloob reviewed (approved):
Nice!
